src/main/java/se/liu/ida/hefquin/engine/query/CypherQuery.java

```diff
@@ -0,0 +1,34 @@
+package se.liu.ida.hefquin.engine.query;
+
+import se.liu.ida.hefquin.engine.query.cypher.CypherVar;
+import se.liu.ida.hefquin.engine.query.cypher.MatchClause;
+import se.liu.ida.hefquin.engine.query.cypher.ReturnStatement;
+import se.liu.ida.hefquin.engine.query.cypher.WhereCondition;
+import se.liu.ida.hefquin.engine.query.impl.MatchCypherQuery;
+import se.liu.ida.hefquin.engine.query.impl.UnionCypherQuery;
+
+import java.util.Set;
+
+public interface CypherQuery {
+    String toString();
+
+    Set<MatchClause> getMatches();
+    Set<WhereCondition> getConditions();
+    Set<ReturnStatement> getReturnExprs();
+    Set<CypherQuery> getUnion();
+    Set<CypherQuery> getIntersect();
```

Comment:
Two comments about this:
1. I understand that, for each of these, the order of the respective elements does not make a difference in terms of query semantics. However, I would consider these classes and interfaces in this PR as elements of ASTs of Cypher queries, and there everything would have an order. Additionally, this order might make a difference in terms of performance. Therefore, I suggest to replace these sets by lists.
2. You are trying to capture three different types of Cypher queries with this one interface; as a consequence, you have a mixture of methods that are not all relevant for all of these types. This is not such a good design. Please create three separate subinterfaces where each of them extends `CypherQuery` and captures only one of the three types (e.g., `CypherMatchQuery`, `CypherUnionQuery`).